File: 1184-car-pooling/car-pooling.py
class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        trips.sort(key = lambda   t : t[1])
        minHeap  = []  # ( END , NUMPASSENGERS )

        curPass = 0

        for numPass , start , end  in trips:
            while minHeap and minHeap[0][0] <= start:
                end1 , numPass1 = heapq.heappop(minHeap)
                curPass -= numPass1
            
            curPass += numPass 

            if curPass > capacity:
                return False 
            
            heapq.heappush(minHeap ,(end , numPass))

        return True 


        

File: 1184-car-pooling/car-pooling.java
import java.util.*;

class Solution {
    public boolean carPooling(int[][] trips, int capacity) {
        Arrays.sort(trips, (a, b) -> Integer.compare(a[1], b[1]));
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        int curPass = 0;

        for (int[] t : trips) {
            int pas = t[0], st = t[1], end = t[2];

            while (!minHeap.isEmpty() && minHeap.peek()[0] <= st) {
                curPass -= minHeap.poll()[1];
            }
            curPass += pas;

            if (curPass > capacity) {
                return false;
            }
            minHeap.offer(new int[]{end, pas});
        }
        return true;
    }
}
